share/doc/wake/tutorial.md

```diff
@@ -29,112 +29,224 @@ The verbose output is of the form `expression: type = value`. The above
 so our new file `tutorial.wake` is available to us from the command-line.
 The syntax `def x = y` introduces a new variable `x` whose value is equal
 to `y`. In this case, `y` is just a `String`.
 
+## Data types and pattern matching
+
+So far, we've gotten a lot done with primitive types (`Integer`, `String`,
+...) and `Lists`. However, wake does allow you to define your own data types.
+These can then be analyzed using pattern matching.
+
+Consider the follow program:
+
+    cat >> tutorial.wake <<'EOF'
+    data Animal =
+        Cat (name: String)
+        Dog (age: Integer)
+
+    def strAnimal a = match a
+        Cat x = "a cat called {x}"
+        Dog y = "a {str y}-year-old dog"
+    EOF
+    wake -vx 'Cat'
+    wake -x 'strAnimal (Dog 12)'
+    wake -x 'strAnimal (Cat "Fluffy")'
+
+The `data` keyword introduces a new type, `Animal`.
+Types are always capitalized, and use a different namespace from variables.
+As defined, `Animal` can either be a `Cat` or a `Dog`,
+where `Cat`s have names and `Dog`s have ages.
+The general syntax is `data TYPE = (CONS TYPE*)+`.
+If we want the new type available to other files, we put an `export` in
+front of the `data`, just like with variables.
+
+As wake informs us, `Cat` is a function that takes a `String` and returns an
+`Animal`. However, unlike normal functions, `Cat` is also a type constructor.
+Type constructors differ from normal variables in that they are capitalized
+and can be used in pattern matches.
+
+The function `strAnimal` is an example of pattern matching. It consists of the
+keyword `match` followed by the value to pattern match. On the following lines
+we indent and provide one or more patterns.
+A pattern always starts with a type constructor followed by a
+number of values corresponding to the values in the constructor declaration.
+These values are followed by `=` and then an expression that is the result
+of when the pattern is matched. For example, in `Cat x = "a cat called {x}"`,
+`x` is a new variable binding corresponding to `name` in the declaration, thus
+it is a `String`. After `=` we have a `String` that is the result whenever `strAnimal`
+is passed a `Cat`.
+
+Note that when matching, the patterns must exhaustively catch all possibilities.
+For example, it would be illegal to `match` but only provide a pattern for `Cat`.
+A useful pattern is to provide a "default" pattern with `_` to match anything:
+
+    def matchWithDefault a = match a
+        Cat x = "a cat called {x}"
+        _     = "not a cat!"
+
+When, as above, we only care about a single constructor of a type, we can use
+the `require` keyword instead. If the object passed to the `require` is of a
+different constructor, the value is passed up a level and the evaluation is
+returned from early (similar to a `return` statement in C-derived languages).
+
+    def addDogYears years pet =
+        require Dog age = pet
+        Dog (age + years * 7)
```

Comment:
I think with imperative backgrounds both the else and non-else cases make equal sense; either way you can phrase it as an early return and that will explain it well. You already made the comparison to the match statement above though so showing an alternative syntax to it is helpful in that explanation. Like literally just translate the above `matchWithDefault` into a require example and explain it the same way is my recommendation. Later you can show that the `else` is optional. The path of ideas is gradual and smooth and all syntax is covered.

I don't think `if` has any role to play here and we should not make the comparison. I think mentioning it will confuse people more than just ignoring it. Not sure though, you should write it both ways and ask Melody which makes more sense and go with that!